### PR TITLE
Billing display very large batches

### PR DESCRIPTION
**Pull request by milo-hyben**
- Restyling Cost By Analysis page.
- Integrating React Virtuoso to display batches with 8K jobs.
- Added BQ cost into HTTP Response header 

Screenshots of the collapsed / expanded records:

Screenshot 2024-03-12 at 4 28 41 PM
Screenshot 2024-03-12 at 4 32 29 PM
Screenshot 2024-03-12 at 4 32 06 PM
Screenshot 2024-03-12 at 4 30 42 PM
Screenshot 2024-03-12 at 4 29 56 PM
Screenshot 2024-03-12 at 4 29 36 PM
Screenshot 2024-03-12 at 4 29 12 PM
Screenshot 2024-03-12 at 4 31 09 PM.31.44 PM.png…]()

### CODE REVIEW

File: db/python/tables/bq/billing_daily_extended.py [hunk 1]
```diff
@@ -49,3 +53,317 @@ async def get_extended_values(self, field: str):
...
+        # getting all records in one go to limit number of queries,
+        # cutting the total cost of BQ
+        # After getting all the needed data,
+        # aggregate:
+        # the cost per topic,
+        # get counts of batches,
+        # cost per hail, cromwell and dataproc
+        # start and end date of the usage
+        # total cost,
+        # including cost by sku
```
Comment (illusional) [inline db/python/tables/bq/billing_daily_extended.py:96]: This text feels a little strangely formatted, maybe needs more indentation if they're list items.

File: db/python/tables/bq/billing_daily_extended.py [hunk 2]
```diff
@@ -49,3 +53,317 @@ async def get_extended_values(self, field: str):
...
+
+        _query = f"""
+            -- get all data we need for aggregations in one go
+            WITH d AS (
```
Comment (illusional) [inline db/python/tables/bq/billing_daily_extended.py:100]: I can't interpret this, chatgpt likes it - mentioning a lot of groupbys may be compute heavy, and

> Consider whether all ORDER BY clauses are necessary, especially in intermediate CTEs, as they can add overhead.

But we pay by bytes processed, and not by compute level - so I'm not overly stressed

File: web/src/pages/billing/BillingCostByAnalysis.tsx [hunk 3]
```diff
@@ -27,6 +29,36 @@ const BillingCostByAnalysis: React.FunctionComponent = () => {
...
+            // nothing found
+            setIsLoading(false)
+            return
+        }
+        const ar_record = arData[0]
+        if (
+            ar_record !== undefined &&
+            ar_record.total !== undefined &&
+            ar_record.total.ar_guid !== undefined
+        ) {
```
Comment (illusional) [inline web/src/pages/billing/BillingCostByAnalysis.tsx:46]: Use optional chaining to simplify this:

```suggestion
        if (!!ar_record?.total?.ar_guid) {
```

File: web/src/pages/billing/BillingCostByAnalysis.tsx [hunk 4]
```diff
@@ -27,6 +29,36 @@ const BillingCostByAnalysis: React.FunctionComponent = () => {
...
+            new AnalysisApi()
+                .getAnalysisRunnerLog(undefined, undefined, ar_record.total.ar_guid, undefined)
```
Comment (illusional) [inline web/src/pages/billing/BillingCostByAnalysis.tsx:48]: Excited to change this haha

File: web/src/pages/billing/BillingCostByAnalysis.tsx [hunk 5]
```diff
@@ -51,7 +83,7 @@ const BillingCostByAnalysis: React.FunctionComponent = () => {
         navigate(url)
     }
 
-    const getData = (sType: SearchType | undefined | string, sTxt: string) => {
+    const getArData = (sType: SearchType | undefined | string, sTxt: string) => {
         if ((sType === undefined || sTxt === undefined) && sTxt.length < 6) {
```
Comment (illusional) [inline web/src/pages/billing/BillingCostByAnalysis.tsx:87]: Wouldn't this fail is sTxt === undefined && sTxt.length < 6 (a null access)?

File: web/src/shared/components/Graphs/DonutChart.tsx [hunk 6]
```diff
@@ -9,10 +9,14 @@ export interface IDonutChartData {
 }
 
 export interface IDonutChartProps {
+    id?: string | undefined
```
Comment (illusional) [inline web/src/shared/components/Graphs/DonutChart.tsx:12]: The question mark implies undefined here (and below)
```suggestion
    id?: string
```

File: web/src/pages/billing/components/HailBatchGrid.tsx [hunk 7]
```diff
@@ -1,258 +1,36 @@
 import * as React from 'react'
```
Comment (illusional) [inline web/src/pages/billing/components/HailBatchGrid.tsx:1]: Do you want to rename this file?

**Review by illusional — APPROVED** [review]
Some changes to address before merging, but otherwise looks good. I think beneficial to release, and see how people find this :)

File: db/python/layers/bq_base.py [hunk 8]
```diff
@@ -11,3 +11,8 @@ def __init__(self, connection: BqConnection):
     def author(self):
         """Get author from connection"""
         return self.connection.author
+
+    @property
+    def cost(self):
+        """Get author from connection"""
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply db/python/layers/bq_base.py:17]: Done

File: db/python/tables/bq/billing_daily_extended.py (hunk 2 above)
Comment (milo-hyben) [reply db/python/tables/bq/billing_daily_extended.py:100]: Yes, bit heavy, but all CTE 's are just groupings, the order by is needed for few as we want to get data pre-order, e.g. cost per topic, sku's in DESC order so no need to reorder on the client side.

File: web/src/pages/billing/BillingCostByAnalysis.tsx (hunk 3 above)
Comment (milo-hyben) [reply web/src/pages/billing/BillingCostByAnalysis.tsx:46]: Thanks for the tip, have not thought about that :-)

File: web/src/shared/components/Graphs/DonutChart.tsx (hunk 6 above)
Comment (milo-hyben) [reply web/src/shared/components/Graphs/DonutChart.tsx:12]: Excellent point :-)

File: web/src/pages/billing/components/HailBatchGrid.tsx (hunk 7 above)
Comment (milo-hyben) [reply web/src/pages/billing/components/HailBatchGrid.tsx:1]: Yes, good point, changed to BathGrid

File: api/routes/billing.py [hunk 9]
```diff
@@ -273,34 +276,38 @@ async def get_namespaces(
...
     ar_guid: str,
     author: str = get_author,
-) -> BillingHailBatchCostRecord:
+) -> JSONResponse:
     """Get Hail Batch costs by AR GUID"""
     billing_layer = _get_billing_layer_from(author)
     records = await billing_layer.get_cost_by_ar_guid(ar_guid)
-    return records
+    headers = {'x-bq-cost': str(billing_layer.connection.cost)}
+    json_compatible_item_data = jsonable_encoder(records)
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply api/routes/billing.py:291]: Yes, well, atm there is no way middleware could access billing_layer.connection.cost. 
Unless I am missing something.
One of the ways would be to make the cost part of the response_model object, but I feel I am polluting response. There is billing cost and then we would have as well query cost.
Maybe we can add something like metadata to each of the response_model and then in the middleware lookup for the metadata.
I would leave it for now so we can get it out, and we can revisit later, if you are fine with that.

**Comment by milo-hyben** [comment]
Hey @illusional , please let me know if you are happy with the latest changes and if you are fine with that one comment.

File: api/routes/billing.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (illusional) [reply api/routes/billing.py:291]: Yeah agreed, let's leave it now, and maybe relook at it later :)

**Review by illusional — APPROVED** [review]

**Pull request merged**